packages/presence-signal/src/createPresenceSignal.ts

```diff
@@ -0,0 +1,108 @@
+import { createSignal, createEffect, onCleanup, type Accessor } from "solid-js";
+
+type SharedTransitionConfig = {
+  /** Duration in milliseconds used both for enter and exit transitions. */
+  transitionDuration: number;
+};
+
+type SeparateTransitionConfig = {
+  /** Duration in milliseconds used for enter transitions (overrides `transitionDuration` if provided). */
+  enterTransitionDuration: number;
+  /** Duration in milliseconds used for exit transitions (overrides `transitionDuration` if provided). */
+  exitTransitionDuration: number;
+};
+
+type Options = (
+  | SharedTransitionConfig
+  | SeparateTransitionConfig
+  | (SharedTransitionConfig & SeparateTransitionConfig)
+) & {
+  /** Opt-in to animating the entering of an element if `isVisible` is `true` during the initial mount. */
+  initialEnter?: boolean;
+};
+
+/**
+ * Animates the appearance of its children.
+ */
+export default function createPresenceSignal(
+  /** Indicates whether the component that the resulting values will be used upon should be visible to the user. */
+  isVisible: Accessor<boolean>,
+  optsValue: Options | Accessor<Options>,
+) {
+  const opts = () => (typeof optsValue === "function" ? optsValue() : optsValue);
+  const exitTransitionDuration = () => {
+    const optsValue = opts();
+    return "exitTransitionDuration" in optsValue
+      ? optsValue.exitTransitionDuration
+      : optsValue.transitionDuration;
+  };
+  const enterTransitionDuration = () => {
+    const optsValue = opts();
+    return "enterTransitionDuration" in optsValue
+      ? optsValue.enterTransitionDuration
+      : optsValue.transitionDuration;
+  };
+
+  const initialEnter = () => opts().initialEnter ?? false;
+  const [animateIsVisible, setAnimateIsVisible] = createSignal(
+    initialEnter() ? false : isVisible(),
+  );
+  const [isMounted, setIsMounted] = createSignal(isVisible());
+  const [hasEntered, setHasEntered] = createSignal(initialEnter() ? false : isVisible());
+
+  const isExiting = () => isMounted() && !isVisible();
+  const isEntering = () => isVisible() && !hasEntered();
+  const isAnimating = () => isEntering() || isExiting();
```

Comment:
So we have individual signals for `isMounted` and `isVisible` - they can be listened to separately and will only trigger change when the value actually changes.
While `isExiting`, `isEntering`, and `isAnimating` is derived from them, it's not done with a memo, so they can trigger updates even if the value didn't actually change.
I don't like the inconsistency here. As a user, if I get an accessor from a custom primitive, my impression would be that the value is memorized. How I'm I supposed to tell which accessors to memorize and which I don't have to?

Reply:
It hadn't actually crossed my mind that these could trigger updates that didn't need to happen, but you're definitely right. Fixed 👍 

Also did the same for the derived `exitDuration` & `enterDuration` signals seeing as they could potentially trigger reactive updates any time any of the options changed.